### ISSUE TITLE
Merge *.model.xml with user's files

### ISSUE THREAD

**Issue by dail8859** [notepad-plus-plus/notepad-plus-plus#1205]
I've seen several times that in order to get updates to keywords, styles, etc, that the user has to delete their `stylers.xml` or `langs.xml` causing them to lose their custom settings. This is because if the `stylers.xml` and `langs.xml` files exist, then they are not modified, even if their `*.model.xml` counter-parts have been updated. 

It would be nice if any new styles or keywords that show up in the `*.model.xml` files are added to the user's files. I would suggest that to begin with this just be overly cautious and just add anything new it finds and not try to do anything like merging lists or modify something that already exists in the user's file. But after making a [few](https://github.com/notepad-plus-plus/notepad-plus-plus/pull/1178) [pull](https://github.com/notepad-plus-plus/notepad-plus-plus/pull/1180) [requests](https://github.com/notepad-plus-plus/notepad-plus-plus/pull/1182) [recently](https://github.com/notepad-plus-plus/notepad-plus-plus/pull/1184) to these model files, it seems like it should be fairly straight forward to see if something new shows up in the model files.

This "merging" wouldn't have to run every time, ideally it would be only when N++ has been updated but I don't know of anything in the code to determine if it is the first time a specific version of N++ has been started. Another option may be to check the modification time of the two files to see if the model files have been updated more recently than the user's local files.

**Comment by BrenySan-DDS** [comment]
Thanks a mill will give it a try...was hoping to avoid the lengthy formatting.. I just hope I can rtecover my usb external files... but so far I Notepad++ rocks...thanks man

**Comment by sanastasiou** [comment]
@dail8859 I could help you with : I don't know of anything in the code to determine if it is the first time a specific version of N++ has been started.

It should be relative easy. We can just save this into the registry or somewhere persistent and check that each time Npp starts. If the versions differ - you got it.

**Comment by dail8859** [comment]
Since this would be a non-trivial patch, I'd like to know it would actually be accepted before spending my limited free time on it.

**Comment by GreenReaper** [comment]
It's perhaps too big a change, but perhaps it'd make more sense to have the user's configuration _override_ the default config, rather than be a modified copy of it? That way any changes in the default automatically take effect unless they've been overridden.

**Comment by frank-orellana** [comment]
**TL;DR:** Don't merge any files. Just backup and replace the **`%appdata%\langs.xml`** file on every installation.

Hi I've had some of this issues and I noticed something.

Almost everything that I customize goes either to the `stylers.xml, userDefineLang.xml or the config.xml` files. BUT the source of all my troubles and the bugs I've seen reported are because of just the **langs.xml** file.

I've searched but i couldn't find a way to modify that file from the options and the only times I've updated it manually it has been to correct the problems of it not being in sync with the **langs.model.xml** file.

If i want a new keyword, new file extension or change to some style of any lang, I change it from the **`Settings -> Style Configurator...`** and **that menu updates `stylers.xml` instead!**, so I can't think of a reason for having that file unless the user really can't do what he wants in the stylers.xml and he really knows what he is doing. 

The only time I can think of we would need this file would be to make very fine modifications to a language (add keywords of different types NOT accessible from the menu for example)but that has not happened to me yet, maybe changing comment char, but that could be added to the menu in my opinion.

I don't remember having needed that file, even when I customize Npp a lot and use a lot of custom languages and make additions to the existing ones.

Even if some user prefers to update manually this file instead of the stylers.xml, he will lose all the new stuff of every new installation. He should update stylers.xml instead, or know he has to backup the file before any new installation, compare it against the new file and merge changes manually (the compare plugin can be easily used in these cases). As It was said, merging automatically is not trivial and maybe it will be made in the future only if it is really needed.

A simple backup of this file in the installation would allow users to have the new stuff, and merge their changes **if the new version does not already provide them**! A warning or question in the installer would prevent any new issues being raised like "I lost my weird changes after installing". Even more, after the new installation is complete, both the backup, if exists, and new langs.xml file could be opened so the user can try to merge changes if any (I would say this is an overkill but it might help some of the users).

Also if the old langs.xml is identical to the langs.model.xml then even a backup should not be needed.

If I delete or rename the file, it is created again with a fresh copy of `langs.model.xml` (And I did not loose any of my custom settings since they were in the other files, and I sure don't miss it).

Stylers.xml is much better for changing the languages because it only has what is changed by the user. We can even improve it later if needed and maybe even get rid of the langs.xml file.

So in short:

1. `%appdata%\langs.xml` is the source of a lot of problems
2. User settings are not stored in that file (even custom keywords and stuff)
3. `%appdata%\langs.xml` can be backed up on installation (if different from `langs.model.xml`)
4. A simple backup could allow users to get all the new stuff and merge their changes if they want
5. `%appdata%\stylers.xml` can be improved in the future to avoid needing this file.

These are some of the open issues that are affected by not updating this file (much more closed):
#2455, #3072 , #780, #3063, #1813, #555, #661,#2640, #1848, etc...

**Comment by ygoe** [comment]
Why is there even a langs.xml file in the user profile directory if the user has nothing to do with it? Just remove it and use the file from the installation directory instead. Problem solved.

And stylers.xml does have *everything* in it, not just what the user has modified. Even worse, what's not in that file can't be modified by the user. This caused me some trouble with an outdated stylers.xml in #4007.

The entire concept of maintaining full copies of files that the user may modify is broken. Instead, only modified data must be stored with the user profile, and files from the installation directory must be used as fallback, or base layer. This is the only method that will survive upgrades without data loss and make new features available to all.

**Comment by LMLB** [comment]
@ygoe The tab settings are stored in langs.xml.

**Comment by MAPJe71** [comment]
In addition: where did `config.model.xml` go?

**Comment by MetaChuh** [comment]
we are closing this issue due to multiple newer notepad++ versions since it was filed.

please update your notepad++ to the latest version [<kbd>Notepad++ 7.6.6</kbd>](https://notepad-plus-plus.org/download/v7.6.6.html) and migrate or reinstall your plugins.
if your issue still persists, and it is still of relevance to you, please feel free to file a new issue as detailed and reproducible as possible.

thank you for your help and best regards.

-----
```
general notes:

please close your issues as soon as you are satisfied, to help
keeping the notepad++ issue tracker clean.

when creating a new issue, please provide all information as shown at the issue template.
your notepad++ debug info is part of the minimum requirements.
(? > Debug Info... > Copy debug info into clipboard).

please also conduct a search at the issue tracker beforehand, to check if a similar issue
already exists.

for general questions, or if it is not sure, whether your issue is directly related to
the notepad++ source code, please visit us at the notepad++ community forum, and
search if related topics exists. you are welcome to post either at similar
topics, or to create a new topic at Help Wanted or General Discussion.
(no extra account is needed, just use your github account to sign in)
```
[<p align="center"><kbd> &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; Click here to visit the Notepad++ Community Forum &emsp; &emsp; &emsp; &emsp; &emsp; &emsp;&emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; </kbd></p>](https://notepad-plus-plus.org/community/)

**Comment by mzso** [comment]
@MetaChuh 
So you're mass closing stuff that no-one bothered to touch, because maybe it was fixed by god in the time that had passed since the issue was opened?
